net/captivedetection/captivedetection.go

```diff
@@ -0,0 +1,189 @@
+	// and to have a short timeout since it will talk to a LAN device.
+	httpClient *http.Client
+}
+
+func NewDetector() *Detector {
+	return &Detector{
+		httpClient: &http.Client{
+			// No redirects allowed
+			CheckRedirect: func(req *http.Request, via []*http.Request) error {
+				return http.ErrUseLastResponse
+			},
+			Transport: http.DefaultClient.Transport,
+			Jar:       http.DefaultClient.Jar,
+			Timeout:   Timeout,
+		},
+	}
+}
+
+// Timeout is the timeout for captive portal detection requests. Because the captive portal intercepting our requests
+// is usually located on the LAN, this is a very short timeout.
+const Timeout time.Duration = 500 * time.Millisecond
+
+// DetectCaptivePortal is the entry point to the API. It attempts to detect if the system is behind a captive portal
+// by making HTTP requests to known captive portal detection Endpoints. If any of the requests return a response code
+// or body that looks like a captive portal, we return true. We return false in all other cases, including when any
+// error occurs during a detection attempt.
+//
+// This function might take a while to return, as it will attempt to detect a captive portal on all available interfaces
+// by performing multiple HTTP requests. It should be called in a separate goroutine if you want to avoid blocking.
+func (d *Detector) DetectCaptivePortal(ifst *netmon.State, derpMap *tailcfg.DERPMap, preferredDERPRegionID int, logf logger.Logf, goos string) (found bool, err error) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	if logf == nil {
+		logf = log.Printf
+	}
+
+	if !ifst.AnyInterfaceUp() {
+		logf("DetectCaptivePortal: no interfaces up, returning false")
+		return false, nil
+	}
+
+	// Here we try detecting a captive portal using *all* available interfaces on the system
+	// that have a IPv4 address. We consider to have found a captive portal when any interface
+	// reports one may exists. This is necessary because most systems have multiple interfaces,
+	// and most importantly on macOS no default route interface is set until the user has accepted
+	// the captive portal alert thrown by the system. If no default route interface is known,
+	// we need to try with anything that might remotely resemble a Wi-Fi interface.
+	for ifName, i := range ifst.Interface {
+		if !i.IsUp() || i.IsLoopback() || interfaceNameDoesNotNeedCaptiveDetection(ifName, goos) {
+			continue
+		}
+		addrs, err := i.Addrs()
+		if err != nil {
+			logf("DetectCaptivePortal: failed to get addresses for interface %s: %v", ifName, err)
+			continue
+		}
+		if len(addrs) == 0 {
+			continue
+		}
+		logf("attempting to do captive portal detection on interface %s", ifName)
+		res := d.detectOnInterface(derpMap, i.Index, preferredDERPRegionID, logf, goos)
+		if res {
+			logf("DetectCaptivePortal(found=true,ifName=%s)", found, ifName)
+			return true, nil
+		}
+	}
+
+	logf("DetectCaptivePortal(found=false)")
+	return false, err
```

Comment:
I don't think we set `err` here; you set it to a value in the `for` loop above, but that's only scoped to that loop. In fact, the only case where we would return an error is if `i.Addrs` returns one, I think?

Reply:
Dropped the error entirely. The API is cleaner if we simply return a bool at all times; the goal of this package is to minimize false positives, so if we hit an error during detection the most reasonable thing is to just tell the caller that we don't believe there's a captive portal.